KIP-699: Update FindCoordinator to resolve multiple Coordinators at a time

mimaison opened the pull request:
This PR implements [KIP-699](https://cwiki.apache.org/confluence/display/KAFKA/KIP-699%3A+Update+FindCoordinator+to+resolve+multiple+Coordinators+at+a+time)

It updates FindCoordinator request and response to support resolving multiple coordinators at a time. If a broker does not support the new FindCoordinator version, clients can revert to the previous behaviour and use a request for each coordinator.

All methods in Admin that require looking up coordinators have been updated to use the new AdminApiDriver logic.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

mimaison committed:
KAFKA-12663: Update FindCoordinator to resolve multiple Coordinators at a time

Changes for KIP-699

skaundinya15 commented:
@mimaison Should be worth noting that there already exists a PR to simplify the `ListOffsets` API here: https://github.com/apache/kafka/pull/10467, so if that gets merged soon, you may need to rebase your implementation on top of that.

mimaison commented:
@skaundinya15 Yeah we may have to rebase but that shouldn't be an issue. I did not touch ListOffsets in this PR, I'm only updating methods that interact with coordinators

mimaison commented:
@rajinisivaram @tombentley @dajac Can you take a look? Thanks

dajac commented:
@mimaison Could you briefly describe the core changes that you have made in the PR? That would be helpful to dive into it. Thanks!

mimaison commented:
@dajac Right, I've updated [the description](https://github.com/apache/kafka/pull/10743#issue-650159295) to give some context

tombentley commented on clients/src/main/java/org/apache/kafka/clients/admin/DeleteConsumerGroupsResult.java:
```diff
@@ -40,7 +43,11 @@
      */
     public Map<String, KafkaFuture<Void>> deletedGroups() {
-        return futures;
+        Map<String, KafkaFuture<Void>> deletedGroups = new HashMap<>();
```
```suggestion
        Map<String, KafkaFuture<Void>> deletedGroups = new HashMap<>(this.futures.size());
```

tombentley commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/AdminApiHandler.java:
```diff
@@ -63,7 +64,7 @@
      * @return result indicating key completion, failure, and unmapping
      */
-    ApiResult<K, V> handleResponse(int brokerId, Set<K> keys, AbstractResponse response);
+    ApiResult<K, V> handleResponse(int brokerId, Set<K> keys, AbstractResponse response, Node node);
```
Needs `@param` javadoc tag

mimaison replied:
I've actually replaced the first argument `int brokerId` by `Node broker` and removed the last argument.

tombentley commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/CoordinatorKey.java:
```diff
@@ -24,7 +24,7 @@
     public final FindCoordinatorRequest.CoordinatorType type;
 
-    private CoordinatorKey(String idValue, FindCoordinatorRequest.CoordinatorType type) {
+    public CoordinatorKey(String idValue, FindCoordinatorRequest.CoordinatorType type) {
```
To me it would be more natural to specify the `type` first then the `idValue`.

mimaison replied:
I agree, since there are very few callers, I'll make the change

tombentley commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/DescribeConsumerGroupsHandler.java:
```diff
@@ -0,0 +1,194 @@
+    private final AdminApiLookupStrategy<CoordinatorKey> lookupStrategy;
+
+    public DescribeConsumerGroupsHandler(
+        Set<String> groupIds,
```
Unused

tombentley commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/AdminApiDriver.java:
```diff
@@ -204,7 +206,8 @@ private void completeLookup(Map<K, Integer> brokerIdMapping) {
         RequestSpec<K> spec,
-        AbstractResponse response
+        AbstractResponse response,
+        Node node
```
Need to fix Javadoc for `#poll()`m which now has a broken link to this method.

mimaison replied:
Good catch!

tombentley commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/AdminApiDriver.java:
```diff
@@ -250,6 +254,13 @@ public void onFailure(
             retryLookup(keysToUnmap);
+
+        } else if (t instanceof UnsupportedBatchLookupException) {
+            ((CoordinatorStrategy) handler.lookupStrategy()).disableBatch();
```
The need for this typecast (i.e. the coupling between the strategy and the exception) makes me think that the name of `UnsupportedBatchLookupException` isn't ideal. It's pretty specific to the `FindCoordinators` API, but that's not really conveyed by its name. Maybe something like `NoBatchedFindCoordinatorsException`?

mimaison replied:
Right `UnsupportedBatchLookupException` is not a great name!

tombentley commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/AdminApiDriver.java:
```diff
@@ -267,7 +278,11 @@ public void onFailure(
         if (requestState != null) {
-            requestState.clearInflight(currentTimeMs);
+            if (spec.scope instanceof FulfillmentScope) {
+                requestState.clearInflight(currentTimeMs + retryBackoffMs);
+            } else {
+                requestState.clearInflight(currentTimeMs);
```
Why the differing behaviour depending on the scope? There's no need to back off on find coordinator requests?

mimaison replied:
That surprised me too. It is the current behaviour that all`*RetryBackoff` tests in `KafkaAdminClientTest` enforce

dajac replied:
I think that this is worth a comment as it is not obvious.

tombentley commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/CoordinatorStrategy.java:
```diff
@@ -40,61 +54,116 @@ public CoordinatorStrategy(
+        keys = keys.stream().filter(k -> isRepresentableKey(k.idValue)).collect(Collectors.toSet());
+        if (batch) {
+            keys = requireSameType(keys);
+            type = keys.iterator().next().type;
```
We can be sure that keys won't be empty, right?

mimaison replied:
Yes `requireSameType` ensures there is 1 type

tombentley commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/CoordinatorStrategy.java:
```diff
@@ -17,20 +17,34 @@
+
     private final Logger log;
+    private boolean batch = true;
+    private FindCoordinatorRequest.CoordinatorType type;
```
Why can't the `type` be set once in the constructor (since all the call sites would appear to know what type they're interested in), rather than needing to be mutable?

mimaison replied:
That's a good idea, yes

tombentley commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:
```diff
@@ -858,6 +885,12 @@ public void onSuccess(ClientResponse resp, RequestFuture<Void> future) {
         public void onFailure(RuntimeException e, RequestFuture<Void> future) {
             log.debug("FindCoordinator request failed due to {}", e.toString());
 
+            if (e instanceof UnsupportedBatchLookupException) {
```
Is there some reason why we can't pass the `ApiVersions` into the coordinator, so we can get the batching right without needing to retry like this?

mimaison replied:
I've only taken a very brief look and I think this approach would work well for Connect, Producer and Consumer, however it's a bit more complicated with Admin.

In Admin, requests are built by lookup strategies. Lookups can be sent to any broker so knowing the max version for a specific call is not completely trivial. That said, it's not impossible either so if there's concensus it would be preferable I can give that a try.

tombentley replied:
This `AdminApiDriver` abstraction is pretty new to me, so I might be wide of the mark, but it doesn't appear to handle this very well. The lookup strategy has to build a request without knowing either the broker or the API versions. It would be possible to pass the `ApiVersions` to the `CoordinatorStrategy`, which should let you do the right thing based on the minimum of the FindCoordinator API version supported in the whole cluster. (That's not completely perfect, since really you'd want to decide on a per-broker basis, but I think it would be good enough). Sadly it's not quite enough to pass just `ApiVersions`, since it doesn't really know about the nodes in the cluster, so you'd need to pass `Metadata` too, which is quite a lot of work. So I can understand why it makes sense to do it like this, since it lets you benefit from the existing logic for figuring out request versions.

skaundinya15 replied:
@tombentley That's a really good point, perhaps we can file a JIRA in the future to add support to the `AdminApiDrive` to support passing in versions for different API calls. I'm currently working on KIP-709 where I'm trying to introduce batching to the fetch offsets API, and I'll need to do something similar as Mickael has done here to be able to support all versions, but having some native support to be able to know the version beforehand would be much more desirable.

tombentley commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1305,66 +1305,102 @@ class KafkaApis(val requestChannel: RequestChannel,
   def handleFindCoordinatorRequest(request: RequestChannel.Request): Unit = {
+    val version = request.header.apiVersion
+    if (version < 4) {
+      handleFindCoordinatorRequestLessThanV4(request)
+    } else {
+      handleFindCoordinatorRequestV4AndAbove(request)
```
Do we really need these separate methods? I would have thought we could just wrap the `findCoordinatorRequest.data.key` in a singleton List for the `version < 4` and otherwise they'd be handled the same. If there is a reason I think it deserves a comment.

mimaison replied:
We need a different `createResponse()` for each version. Both methods call `getCoordinator()` and have no logic.

tombentley replied:
Yes, obvious now you point it out! I guess I can't see a better way of abstracting over that.

tombentley reviewed (commented):
Made a first pass with a few questions.

mimaison committed:
Address feedback

skaundinya15 commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/AlterConsumerGroupOffsetsHandler.java:
```diff
@@ -0,0 +1,156 @@
+    private void handleError(CoordinatorKey groupId, Errors error, Map<CoordinatorKey, Throwable> failed, List<CoordinatorKey> unmapped) {
+        switch (error) {
+            case GROUP_AUTHORIZATION_FAILED:
+                log.error("Received authorization failure for group {} in `DeleteConsumerGroupOffsets` response", groupId,
+                        error.exception());
+                failed.put(groupId, error.exception());
+                break;
+            case COORDINATOR_LOAD_IN_PROGRESS:
+            case COORDINATOR_NOT_AVAILABLE:
+            case NOT_COORDINATOR:
+                log.debug("DeleteConsumerGroupOffsets request for group {} returned error {}. Will retry", groupId, error);
+                unmapped.add(groupId);
+                break;
+            default:
+                log.error("Received unexpected error for group {} in `DeleteConsumerGroupOffsets` response", 
+                        groupId, error.exception());
+                failed.put(groupId, error.exception(
+                        "Unexpected error during DeleteConsumerGroupOffsets lookup for " + groupId));
```
Seems like there's a copy paste error, should be `AlterConsumerGroupOffsets`. Also, when I compare to the `handleError()` in `DescribeConsumerGroupsHandler`, it seems to be slightly different:

```java
    private void handleError(CoordinatorKey groupId, Errors error, Map<CoordinatorKey, Throwable> failed, List<CoordinatorKey> unmapped) {
        switch (error) {
            case GROUP_AUTHORIZATION_FAILED:
                log.error("Received authorization failure for group {} in `DescribeGroups` response", groupId,
                        error.exception());
                failed.put(groupId, error.exception());
                break;
            case COORDINATOR_LOAD_IN_PROGRESS:
            case COORDINATOR_NOT_AVAILABLE:
                break;
            case NOT_COORDINATOR:
                log.debug("DescribeGroups request for group {} returned error {}. Will retry",
                        groupId, error);
                unmapped.add(groupId);
                break;
            default:
                log.error("Received unexpected error for group {} in `DescribeGroups` response", 
                        groupId, error.exception());
                failed.put(groupId, error.exception(
                        "Unexpected error during DescribeGroups lookup for " + groupId));
        }
    }
```

In the `AlterConsumerGroupOffsetsHandler` we don't `break` on `COORDINATOR_NOT_AVAILABLE`, but we do in `DescribeConsumerGroupsHandler` - any reason for this? I'd imagine the error handling across all the consumer groups handler would be the same. Perhaps we could factor this out and put it in some generic `ConsumerGroupHandler` class that implements this, and takes in a request name so it can be used across all the consumer groups handlers. What do you think?

mimaison replied:
Yes some of these handlers different in pretty subtle ways. Unfortunately if we remove this `break` statement, some of the existing tests stop working. In this PR, I've aimed at not touching the exiting tests to ensure my refactoring don't change the current behaviour.

I agree, this is very confusing and I think we should align behaviour for all these similar calls and have a generic error handling logic. But I would prefer to not do it as part of this PR as it's already pretty big.

skaundinya15 replied:
@mimaison makes sense, it would be good to file a JIRA for this so we can address it in a future PR to ensure we have consistent error handling across all consumer group related issues.

mimaison replied:
I agree, I raised https://issues.apache.org/jira/browse/KAFKA-13012

skaundinya15 commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/CoordinatorKey.java:
```diff
@@ -24,7 +24,7 @@
     public final String idValue;
     public final FindCoordinatorRequest.CoordinatorType type;
 
-    private CoordinatorKey(String idValue, FindCoordinatorRequest.CoordinatorType type) {
```
Any reason for switching around the order of `idValue` and `type`?

mimaison replied:
Not a strong reason but [like Tom](https://github.com/apache/kafka/pull/10743#discussion_r646599727), I find the order type, value more natural

skaundinya15 commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/CoordinatorStrategy.java:
```diff
@@ -17,84 +17,153 @@
+        if (keys.stream().map(k -> k.type).collect(Collectors.toSet()).size() != 1) {
+            throw new IllegalArgumentException("Unexpected lookup key set");
+        }
+        return keys;
```
Do we need to return back the `keys`? It seems like this is just used as a sanity check to ensure that all the `CoordinatorKeys` are of the same type. If we don't pass the if statement check, it would just return the `IllegalArgumentException`, otherwise nothing else happens. I ask since it doesn't seem like this method really does anything different to the passed in keys to actually return anything. If we change it to this, then we could change the method name to `ensureSameType`.

skaundinya15 commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/CoordinatorStrategy.java:
```diff
@@ -17,84 +17,153 @@
-                .setKeyType(key.type.id())
-        );
+        unrepresentableKeys = keys.stream().filter(k -> !isRepresentableKey(k.idValue)).collect(Collectors.toSet());
+        keys = keys.stream().filter(k -> isRepresentableKey(k.idValue)).collect(Collectors.toSet());
```
nit: Personally I think I'd prefer if we separated it out to `unrepresentableKeys` and `representableKeys` rather than modify the passed in parameter, but upto you if you want to change that.

skaundinya15 commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/CoordinatorStrategy.java:
```diff
@@ -17,84 +17,153 @@
-
+                mappedKeys.put(key, nodeId);
+                break;
             case COORDINATOR_NOT_AVAILABLE:
```
Would we consider something like `COORDINATOR_NOT_AVAILABLE` a transient exception that we should retry on? I'm a bit confused since we don't log anything or do a `break` on this case. In either case, I think we should log + put it in the `failedKeys` map if we think that that kind of exception should be considered a failure, or let it be if we think it should be a more transient error.

mimaison replied:
This is the current behaviour, when hitting these errors, we retry the find coordinator request. The way to signal we want to retry is to omit the key in the return value of `handleResponse()`, see https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/admin/internals/AdminApiLookupStrategy.java#L71-L74

So I think we want to keep it as it is.

skaundinya15 replied:
@mimaison Gotcha, makes sense. Can we still we can still log at `DEBUG` level? In case we need to debug, we have evidence of this in the logs.

skaundinya15 commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/CoordinatorStrategy.java:
```diff
@@ -17,84 +17,153 @@
+
+    private static CoordinatorKey requireSingleton(Set<CoordinatorKey> keys) {
+        if (keys.size() != 1) {
+            throw new IllegalArgumentException("Unexpected lookup key set");
```
Could we try to have a more descriptive exception? Perhaps something like `Unexpected size of key set: expected 1, but got keys.size()` or something better phrased than that.

skaundinya15 commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/CoordinatorStrategy.java:
```diff
@@ -17,84 +17,153 @@
+        } else {
+            CoordinatorKey key = requireSingleton(keys);
+            return new FindCoordinatorRequest.Builder(
+                new FindCoordinatorRequestData()
+                    .setKey(key.idValue)
+                    .setKeyType(key.type.id())
```
Just for my understanding, we are reverting to the old request format here in case the broker is an older version than the `AdminClient` right?

mimaison replied:
Yes that's right, this is for request version < 4

skaundinya15 commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/DeleteConsumerGroupsHandler.java:
```diff
@@ -0,0 +1,131 @@
+
+    @Override
+    public DeleteGroupsRequest.Builder buildRequest(int brokerId, Set<CoordinatorKey> keys) {
+        List<String> groupIds = keys.stream().map(key -> {
+            if (key.type != FindCoordinatorRequest.CoordinatorType.GROUP) {
+                throw new IllegalArgumentException("Invalid transaction coordinator key " + key +
+                    " when building `DeleteGroups` request");
```
As I mentioned earlier about having a generic `ConsumerGroupHandler` class, we could add this validation in this too, and make sure we call it every time before we build the request. I didn't see this validation being done in some of the handler classes, but it seems like something we should add before every request we build that has to do with consumer groups and looking up group coordinator keys, even if it's just related to one group.

mimaison replied:
Good point! I've actually moved this check into `CoordinatorStrategy` as it's built with the type of CoordinatorKeys it will handle.

skaundinya15 replied:
@mimaison Sounds good, thanks!

skaundinya15 commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:
```diff
@@ -813,34 +815,56 @@ public void handle(SyncGroupResponse syncResponse,
             FindCoordinatorResponse findCoordinatorResponse = (FindCoordinatorResponse) resp.responseBody();
-            Errors error = findCoordinatorResponse.error();
+            if (batch && findCoordinatorResponse.data().coordinators().size() != 1) {
+                log.error("Group coordinator lookup failed: Invalid response");
+                future.raise(new IllegalStateException("Group coordinator lookup failed: Invalid response"));
```
Could we have a more descriptive error message/exception here? Something more about how any coordinator size that's not 1 is unexpected since we are only looking up the coordinator for one group.

skaundinya15 commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java:
```diff
@@ -1524,11 +1534,29 @@ String coordinatorKey() {
             CoordinatorType coordinatorType = CoordinatorType.forId(builder.data().keyType());
 
+            if (batchFindCoordinator && findCoordinatorResponse.data().coordinators().size() != 1) {
+                log.error("Group coordinator lookup failed: Invalid response");
+                fatalError(new IllegalStateException("Group coordinator lookup failed: Invalid response"));
```
Could we have a more descriptive error message/exception here? Something more about how any coordinator size that's not 1 is unexpected since we are only looking up the coordinator for one group.

skaundinya15 commented on clients/src/main/resources/common/message/FindCoordinatorRequest.json:
```diff
@@ -23,12 +23,16 @@
   "flexibleVersions": "3+",
   "fields": [
-    { "name": "Key", "type": "string", "versions": "0+",
+    { "name": "Key", "type": "string", "versions": "0-3",
```
Just for my understanding, do we have to keep the old field here to be able to serve requests from older versioned clients?

mimaison replied:
Yes we have to keep the existing fields to handle all versions

skaundinya15 reviewed (commented):
Thanks for the PR @mimaison, fantastic job with the refactor - makes so much of the code much more readable! I made a pass through the non test files and left some comments. Will take another pass at the test files soon.

mimaison committed:
Address feedback

mimaison committed:
Address feedback + new tests

mimaison commented:
@skaundinya15 Thanks for the review. I believe I've addressed all your comments.

tombentley commented on clients/src/main/java/org/apache/kafka/clients/admin/DeleteConsumerGroupOffsetsResult.java:
```diff
@@ -32,11 +32,11 @@
 
 
-    DeleteConsumerGroupOffsetsResult(KafkaFuture<Map<TopicPartition, Errors>> future, Set<TopicPartition> partitions) {
+    DeleteConsumerGroupOffsetsResult(KafkaFutureImpl<Map<TopicPartition, Errors>> future, Set<TopicPartition> partitions) {
```
I don't think these changes are needed now.

tombentley commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/AdminApiHandler.java:
```diff
@@ -57,13 +58,13 @@
      * result. They will be retried automatically.
      *
-     * @param brokerId the brokerId that the associated request was sent to
+     * @param broker the broker that the associated request was sent to
```
Is it necessarily a broker, or could it be a kraft controller?

mimaison replied:
Not entirely sure about the new naming in place now, but does that still count as a broker?

tombentley replied:
Well `process.roles` is documented like this in `config/kraft/README.md`:

* If `process.roles` is set to `broker`, the server acts as a broker in KRaft mode.
* If `process.roles` is set to `controller`, the server acts as a controller in KRaft mode.
* If `process.roles` is set to `broker,controller`, the server acts as both a broker and a controller in KRaft mode.

which suggests to me that we're using the term "broker"  to mean "a thing which handles Produce and Fetch etc". (However, "controller" is a bit confusing there, since while we might have several servers in the "controller" role only one will be  _the_ controller at any one time. I'm not aware of a good term for "server that is participating in the raft cluster, but might not be the current controller right now").

mimaison replied:
Ok, thanks for the clarifications. I feel like `broker` is still fine here. Otherwise, maybe `node`?

tombentley commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/AlterConsumerGroupOffsetsHandler.java:
```diff
@@ -0,0 +1,156 @@
+import org.apache.kafka.common.utils.LogContext;
+import org.slf4j.Logger;
+
+public class AlterConsumerGroupOffsetsHandler implements AdminApiHandler<CoordinatorKey, Map<TopicPartition, Errors>> {
```
Perhaps it's worth a line of doc to explain: "Sends the OffsetCommitRequest to each group coordinator based on the discovered coordinators for the relevant groups, and handles the results."

skaundinya15 replied:
@tombentley I think the `offsetCommitRequest` only takes in one group at a time, at least that's what it seems like from the `Admin.java` interface: https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/admin/Admin.java#L1145

tombentley commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/AlterConsumerGroupOffsetsHandler.java:
```diff
@@ -0,0 +1,156 @@
+    }
+
+    @Override
+    public OffsetCommitRequest.Builder buildRequest(int brokerId, Set<CoordinatorKey> keys) {
```
brokerId -> coordinatorId, to emphasize the point that we've found the coordinators at this point?

tombentley commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/AlterConsumerGroupOffsetsHandler.java:
```diff
@@ -0,0 +1,156 @@
+                        .setPartitionIndex(entry.getKey().partition());
+                value.add(partition);
+                return value;
+            });
```
Wouldn't this be clearer as a computeIfAbsent()?
```suggestion
            OffsetCommitRequestPartition partition = new OffsetCommitRequestPartition()
                    .setCommittedOffset(oam.offset())
                    .setCommittedLeaderEpoch(oam.leaderEpoch().orElse(-1))
                    .setCommittedMetadata(oam.metadata())
                    .setPartitionIndex(entry.getKey().partition());
            offsetData.computeIfAbsent(topic, key -> new ArrayList<>()).add(partition);
```

tombentley commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/AlterConsumerGroupOffsetsHandler.java:
```diff
@@ -0,0 +1,156 @@
+    }
+
+    @Override
+    public ApiResult<CoordinatorKey, Map<TopicPartition, Errors>> handleResponse(Node broker, Set<CoordinatorKey> groupIds,
```
Again, wdyt of `broker` -> `coordinator`?

tombentley commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/DeleteConsumerGroupOffsetsHandler.java:
```diff
@@ -0,0 +1,145 @@
+import org.apache.kafka.common.utils.LogContext;
+import org.slf4j.Logger;
+
+public class DeleteConsumerGroupOffsetsHandler implements AdminApiHandler<CoordinatorKey, Map<TopicPartition, Errors>> {
```
To save me repeating what I said before about a line of Javadoc and param name -> coordinator, on this and the other AdminApiHandler subclasses?

tombentley commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/DeleteConsumerGroupOffsetsHandler.java:
```diff
@@ -0,0 +1,145 @@
+            response.data().topics().forEach(topic -> 
+                topic.partitions().forEach(partition -> {
+                    Errors partitionError = Errors.forCode(partition.errorCode());
+                    boolean hasError = handleError(groupId, partitionError, failed, unmapped);
```
inline?

tombentley commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/DeleteConsumerGroupsHandler.java:
```diff
@@ -0,0 +1,124 @@
+    public DeleteGroupsRequest.Builder buildRequest(int brokerId, Set<CoordinatorKey> keys) {
+        List<String> groupIds = keys.stream().map(key -> key.idValue).collect(Collectors.toList());
+        DeleteGroupsRequestData data = new DeleteGroupsRequestData()
+                .setGroupsNames(new ArrayList<>(groupIds));
```
Is the copy needed?

tombentley commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:
```diff
@@ -813,34 +815,56 @@ public void handle(SyncGroupResponse syncResponse,
             if (error == Errors.NONE) {
                 synchronized (AbstractCoordinator.this) {
+                    int nodeId = batch
+                            ? findCoordinatorResponse.data().coordinators().get(0).nodeId()
```
Factor out var for `findCoordinatorResponse.data().coordinators().get(0)`?

mimaison replied:
This variable would only exist if `batch` is true which makes it pretty awkward.

tombentley replied:
Sorriy, I think I made an erroneous point like this before. It sucks that there's no good abstraction across the common parts of `FindCoordinatorResponseData.Coordinator` and `FindCoordinatorResponseData`.

tombentley commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java:
```diff
@@ -1524,11 +1534,29 @@ String coordinatorKey() {
+                fatalError(new IllegalStateException("Group coordinator lookup failed: Invalid response containing more than a single coordinator"));
+            }
+            String key = batchFindCoordinator
+                    ? findCoordinatorResponse.data().coordinators().get(0).key()
```
Similarly a var for `findCoordinatorResponse.data().coordinators().get(0)`

tombentley commented on clients/src/test/java/org/apache/kafka/clients/MockClient.java:
```diff
@@ -245,10 +246,17 @@ public void send(ClientRequest request, long now) {
+                try {
+                    AbstractRequest abstractRequest = request.requestBuilder().build(version);
+                    if (!futureResp.requestMatcher.matches(abstractRequest))
+                        continue;
```
Is it right that we've lost the ISE entirely here?

mimaison replied:
Oops! Yes I removed this to silence failing tests while refactoring. I'll undo this.

tombentley reviewed (commented):
This is looking pretty good a few more nits and questions.

mimaison committed:
Address feedback

mimaison commented:
@tombentley Thanks for the review! I've addressed your findings

tombentley reviewed (approved):
LGTM

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/ConsumerGroupDescription.java:
```diff
@@ -49,7 +49,7 @@ public ConsumerGroupDescription(String groupId,
     }
 
-    ConsumerGroupDescription(String groupId,
+    public ConsumerGroupDescription(String groupId,
```
Could we realign the arguments?

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/AlterConsumerGroupOffsetsHandler.java:
```diff
@@ -0,0 +1,150 @@
+    }
+
+    public static AdminApiFuture.SimpleAdminApiFuture<CoordinatorKey, Map<TopicPartition, Errors>> newFuture(
+            String groupId
```
The indentation of the arguments is not consistent in few places. For instance here, it should be consistent with the constructor.

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/AlterConsumerGroupOffsetsHandler.java:
```diff
@@ -0,0 +1,150 @@
+
+    @Override
+    public ApiResult<CoordinatorKey, Map<TopicPartition, Errors>> handleResponse(Node coordinator, Set<CoordinatorKey> groupIds,
+            AbstractResponse abstractResponse) {
```
Should we adopt one of the other formatting style already used in the PR?

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/ListConsumerGroupOffsetsHandler.java:
```diff
@@ -0,0 +1,142 @@
+
+    @Override
+    public ApiResult<CoordinatorKey, Map<TopicPartition, OffsetAndMetadata>> handleResponse(
+            Node coordinator,
```
Indentation is not consistent here.

dajac reviewed (commented):
Thanks for the PR. I have made a pass over it. Overall, it looks good. A small suggestion: It would be great if we could unify the formatting style. For instance, the style and the indentation of the method declarations is not consistent accros the board.

mimaison committed:
Fix formatting style

mimaison committed:
More formatting fixes

mimaison commented:
Thanks @dajac for the review, I've addressed your comments

skaundinya15 commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/ListConsumerGroupOffsetsHandler.java:
```diff
@@ -0,0 +1,148 @@
+    ) {
+        switch (error) {
+            case GROUP_AUTHORIZATION_FAILED:
+                log.error("Received authorization failure for group {} in `DeleteConsumerGroupOffsets` response", groupId,
```
Seems like there's a copy/paste error here. I think we want `DeleteConsumerGroupOffsets` here to be `ListConsumerGroupOffsets`. Perhaps instead of writing the name for each one of these, we could just call `apiName()` instead?

skaundinya15 commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/RemoveMembersFromConsumerGroupHandler.java:
```diff
@@ -0,0 +1,133 @@
+    ) {
+        switch (error) {
+            case GROUP_AUTHORIZATION_FAILED:
+                log.error("Received authorization failure for group {} in `DeleteConsumerGroupOffsets` response", groupId,
```
Another place where I think we want this to be `RemoveMembersFromConsumerGroup` instead of `DeleteConsumerGroupOffsets`

skaundinya15 reviewed (approved):
Thanks for the updates @mimaison! Just left a couple minor comments, other than that LGTM! Really excited for this patch to make it in :)

skaundinya15 commented on clients/src/main/resources/common/message/FindCoordinatorRequest.json:
```diff
@@ -23,12 +23,16 @@
-    { "name": "Key", "type": "string", "versions": "0+",
+    { "name": "Key", "type": "string", "versions": "0-3",
       "about": "The coordinator key." },
     { "name": "KeyType", "type": "int8", "versions": "1+", "default": "0", "ignorable": false,
```
Should we also change the `versions` for this to `0-3` rather than `1+`?

mimaison replied:
`KeyType` was added in version 1 to resolve transaction coordinators. Before that, FindCoordinator was only able to resolve group coordinators.
It's still used in version 4 as we typically never want to resolve transaction and consumer coordinators at the same time so it determines the type of all keys in the request.

skaundinya15 replied:
Thanks for the explanation @mimaison, makes sense.

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/DeleteConsumerGroupsResult.java:
```diff
@@ -29,9 +32,9 @@
+    private final Map<CoordinatorKey, KafkaFutureImpl<Void>> futures;
 
-    DeleteConsumerGroupsResult(final Map<String, KafkaFuture<Void>> futures) {
+    DeleteConsumerGroupsResult(Map<CoordinatorKey, KafkaFutureImpl<Void>> futures) {
```
Do we really need `KafkaFutureImpl` here? It would be better to use `KafkaFuture` if we can.

skaundinya15 replied:
@dajac @mimaison It looks like the reason we have to do `KafkaFutureImpl` here is because in the `SimpleAdminApiFuture` class, we have a class variable called `futures` which is of type ` private final Map<K, KafkaFutureImpl<V>>`. As a result, we return `KafkaFutureImpl` for the `all()`, `get()` and other methods. We could change all of this to use `KafkaFuture` instead, but this would require us to change the following methods from `protected` to `public`:

```java
    /**
     * If not already completed, sets the value returned by get() and related methods to the given
     * value.
     */
    protected abstract boolean complete(T newValue);

    /**
     * If not already completed, causes invocations of get() and related methods to throw the given
     * exception.
     */
    protected abstract boolean completeExceptionally(Throwable newException);
```

The question here would be should we change these `protected` methods to be `public` so we can maintain just returning a type of `KafkaFuture` or if it's okay to return `KafkaFutureImpl`. I think it could be worth changing `KafkaFuture` methods from `protected` to `public` so that we can maintain the invariant of always returning the type of `KafkaFuture`, but not sure if that would require getting some consensus on the mailing list about this.

tombentley replied:
@skaundinya15 those methods in `KafkaFuture` are intentionally not `public` because a user receiving an instance should _never_ need to complete the future (they're always completed by the admin client). Making them `public` would thus make the API less type safe.

skaundinya15 replied:
@tombentley I see, thank you for the clarification. I guess in that case we have to stick with returning the `KafkaFutureImpl` in this situation then? Doesn't really seem to be a way around this otherwise.

mimaison replied:
Yes that's exactly the reason why I made this change. I'll add the `final` modifier back though

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/DeleteConsumerGroupsResult.java:
```diff
@@ -40,7 +43,11 @@
+        Map<String, KafkaFuture<Void>> deletedGroups = new HashMap<>(futures.size());
+        for (Map.Entry<CoordinatorKey, KafkaFutureImpl<Void>> entry : futures.entrySet()) {
+            deletedGroups.put(entry.getKey().idValue, entry.getValue());
+        }
```
nit: We tend to use `forEach` nowadays which makes the code a bit less verbose. Up to you.

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/DescribeConsumerGroupsResult.java:
```diff
@@ -34,39 +35,40 @@
+    private final Map<CoordinatorKey, KafkaFutureImpl<ConsumerGroupDescription>> futures;
 
-    public DescribeConsumerGroupsResult(final Map<String, KafkaFuture<ConsumerGroupDescription>> futures) {
+    public DescribeConsumerGroupsResult(Map<CoordinatorKey, KafkaFutureImpl<ConsumerGroupDescription>> futures) {
```
Same question regarding `KafkaFutureImpl`.

mimaison replied:
Explained in https://github.com/apache/kafka/pull/10743#discussion_r657473433

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/CoordinatorKey.java:
```diff
@@ -24,7 +24,7 @@
     public final FindCoordinatorRequest.CoordinatorType type;
 
-    private CoordinatorKey(String idValue, FindCoordinatorRequest.CoordinatorType type) {
+    public CoordinatorKey(FindCoordinatorRequest.CoordinatorType type, String idValue) {
```
nit: Could we keep this private? I think that the intent was to use the static methods below to construct it.

mimaison replied:
I was using it in `CoordinatorStrategy.handleResponse()` but I get easily get around it. It makes sense to keep it private

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/CoordinatorStrategy.java:
```diff
@@ -17,84 +17,160 @@
+                return GROUP_REQUEST_SCOPE;
+            } else {
+                return TXN_REQUEST_SCOPE;
+            }
```
nit: I wonder if differentiating the two is required here. It seems that we don't use them anywhere else so we could use one common `REQUEST_SCOPE` for both cases.

mimaison replied:
Right, I merged both as `BATCH_REQUEST_SCOPE`

skaundinya15 commented on clients/src/main/java/org/apache/kafka/common/requests/FindCoordinatorRequest.java:
```diff
@@ -70,7 +75,7 @@ public AbstractResponse getErrorResponse(int throttleTimeMs, Throwable e) {
         }
         Errors error = Errors.forException(e);
-        return FindCoordinatorResponse.prepareResponse(error, Node.noNode());
+        return FindCoordinatorResponse.prepareOldResponse(error, Node.noNode());
```
Why do we call the `prepareOldResponse` in `getErrorResponse`? Should we be doing a version check here?

dajac commented on clients/src/main/java/org/apache/kafka/common/errors/NoBatchedFindCoordinatorsException.java:
```diff
@@ -0,0 +1,33 @@
+ * Indicates that it is not possible to lookup coordinators in batches with FindCoordinator. Instead
+ * coordinators must be looked up one by one.
+ */
+public class NoBatchedFindCoordinatorsException extends UnsupportedVersionException {
```
If I remember correctly, this new exception was not specified in the KIP. Should we update it and notify the thread in the mailing list?

mimaison replied:
Right, I'll update the KIP and thread

tombentley replied:
It seems like a mistake to add this to the public `errors` package when it should never propagate outside the client. An alternative could to be make it a static member class of `FindCoordinatorRequest`, or put it in some other package.

dajac replied:
Yeah, I was also thinking about this. It is definitely better if we could keep it private.

mimaison replied:
I agree, I'll move it to an inner class of `FindCoordinatorRequest` and update the KIP accordingly

dajac replied:
We need to discuss this further.

mimaison replied:
I pushed a change that moves `NoBatchedFindCoordinatorsException` into `FindCoordinatorRequest`. Do you have any further concerns?

dajac replied:
Nope, that's fine. Thanks.

skaundinya15 commented on clients/src/main/java/org/apache/kafka/common/requests/FindCoordinatorResponse.java:
```diff
@@ -95,4 +96,21 @@ public static FindCoordinatorResponse prepareResponse(Errors error, Node node) {
+        return new FindCoordinatorResponse(data);
+    }
+
+    public static boolean isBatch(RequestHeader header) {
+        return header.apiVersion() >= 4;
+    }
```
This doesn't seem to be used anywhere, is this supposed to be there for convenience for the future?

mimaison replied:
Right, deleted

dajac commented on clients/src/test/java/org/apache/kafka/clients/MockClient.java:
```diff
@@ -245,10 +246,18 @@ public void send(ClientRequest request, long now) {
+                    if (!futureResp.requestMatcher.matches(abstractRequest))
+                        throw new IllegalStateException("Request matcher did not match next-in-line request "
+                                + abstractRequest + " with prepared response " + futureResp.responseBody);
+                } catch (NoBatchedFindCoordinatorsException uble) {
```
I wonder if we should catch `UnsupportedVersionException` here to handle the general case. Have you considered it? Also, the name `uble` looks weird. Is it intentional?

mimaison replied:
Yes I looked into it but I think there's 1 test in TransactionManagerTest that expects `UnsupportedVersionException` to be thrown.
MockClient works in slightly different ways than the real client and it would be good to address this but I'd rather defer to a follow up PR.

dajac replied:
Understood. Should we file a Jira about this?

mimaison replied:
Sure, I raised https://issues.apache.org/jira/browse/KAFKA-13000

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -4806,7 +4275,7 @@ public ListTransactionsResult listTransactions(ListTransactionsOptions options)
             void handleResponse(AbstractResponse response) {
                 long currentTimeMs = time.milliseconds();
-                driver.onResponse(currentTimeMs, spec, response);
+                driver.onResponse(currentTimeMs, spec, response, nodeProvider.provide());
```
Does getting the node with `nodeProvider.provide()` work in all cases? I suppose that it does not guarantee returning the correct node if `LeastLoadedNodeProvider` is used. Am I getting this right?

Looking at the code, I am not sure to understand why/where we need this. Could you elaborate a bit?

mimaison replied:
All of this is specifically because `ConsumerGroupDescription` needs the coordinator as a `Node` when describing consumer groups.

At the moment, handlers only get the brokerId and that's not enough to build a `Node` object. So instead in this PR, I replaced the `brokerId` argument by the full `Node` object. The `brokerId` can be retrieved from `spec.scope` but for the `Node`, we need to get that from the `Call` itself. 

Good catch, `nodeProvider.provide()` could return a different `Node`! As far as I can tell, we can use `curNode()` here, that should have been set by `maybeDrainPendingCall()`.

dajac replied:
I see. That seems reasonable to me now.

dajac reviewed (commented):
@mimaison Thanks for the updates. I left a few more comments/questions.

skaundinya15 commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/DeleteConsumerGroupsHandler.java:
```diff
@@ -0,0 +1,135 @@
+    private final AdminApiLookupStrategy<CoordinatorKey> lookupStrategy;
+
+    public DeleteConsumerGroupsHandler(
+        Set<String> groupIds,
```
This parameter passed into the constructor doesn't seem to be used.

skaundinya15 commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/CoordinatorKey.java:
```diff
@@ -24,7 +24,7 @@
     public final String idValue;
```
Since we are now going to be returning a `CoordinatorKey` type in some of the results we return (eg: `DescribeConsumerGroupResult`, `DeleteConsumerGroupResult`), perhaps we should move this out of the the `internals` package and instead put it in the `clients/src/main/java/org/apache/kafka/clients/admin` directory instead.

mimaison replied:
This package is part of the public API. I'm not sure we want to expose `CoordinatorKey` which is currently only used internally.
That said, [KIP-692](https://cwiki.apache.org/confluence/display/KAFKA/KIP-692%3A+Make+AdminClient+value+object+constructors+public) proposes exposing the constructors of all the `*Result` classes. Maybe that discussion should happen there?

skaundinya15 replied:
Yeah I can definitely propose that in the discussion thread there. As I was working on my implementation on KIP-709 for batching the `listConsumerGroupsOffsets` API, I saw the need to make this class public since it will be needed to handle the `ListConsumerGroupsOffsetsResult` class properly. But I'll take this discussion to the email thread/the PR I'll open shortly.

dajac replied:
@skaundinya15 I don't understand why we would want to return `CoordinatorKey` in some results. Do we really have to? It seems that we could unwrap it depending on the context and only expose what the user need (e.g. group id).

Personally, I believe that we should strive for keeping all the constructors of `*Result` classes package private for the time being until KIP-692 is accepted, if it will.

skaundinya15 replied:
@mimaison @dajac Ah I see the issue now, you're right - we don't need to return `CoordinatorKey` as part of the result. I was mistaken here - we can just keep the original method signature and convert the internal `CoordinatorKey` mapping to the `String` group id. Apologies for the confusion, the way that @mimaison has it here works just fine, I'll use a similar approach when implementing KIP-709.

mimaison committed:
Address feedback

mimaison committed:
Address comments

mimaison commented:
@tombentley @dajac @skaundinya15 Thanks for the reviews! I believe I've addressed all your comments now.

skaundinya15 commented on clients/src/main/java/org/apache/kafka/clients/admin/DescribeConsumerGroupsResult.java:
```diff
@@ -34,39 +35,40 @@
     public Map<String, KafkaFuture<ConsumerGroupDescription>> describedGroups() {
-        return futures;
+        Map<String, KafkaFuture<ConsumerGroupDescription>> describedGroups = new HashMap<>();
+        for (Map.Entry<CoordinatorKey, KafkaFutureImpl<ConsumerGroupDescription>> entry : futures.entrySet()) {
+            describedGroups.put(entry.getKey().idValue, entry.getValue());
+        }
```
Nit: Similar to how it's done in `DeleteConsumerGroupsResult`, perhaps we can replace this with a lambda expression instead?
```java
futures.forEach((key, future) -> describedGroups.put(key.idValue, future));
```

skaundinya15 commented on clients/src/main/java/org/apache/kafka/clients/admin/DeleteConsumerGroupsResult.java:
```diff
@@ -40,7 +43,11 @@
     public Map<String, KafkaFuture<Void>> deletedGroups() {
-        return futures;
+        Map<String, KafkaFuture<Void>> deletedGroups = new HashMap<>(futures.size());
+        futures.forEach((key, future) -> {
+            deletedGroups.put(key.idValue, future);
+        });
```
Nit: I think we could simplify this lambda expression further to just:

```java
futures.forEach((key, future) -> deletedGroups.put(key.idValue, future));
```

skaundinya15 commented on clients/src/main/java/org/apache/kafka/clients/admin/DescribeConsumerGroupsResult.java:
```diff
@@ -34,39 +35,40 @@
+            nil -> {
+                Map<String, ConsumerGroupDescription> descriptions = new HashMap<>(futures.size());
+                try {
+                    for (Map.Entry<CoordinatorKey, KafkaFutureImpl<ConsumerGroupDescription>> entry : futures.entrySet()) {
```
Nit: We could also use the lambda expression instead:

```java
    public KafkaFuture<Map<String, ConsumerGroupDescription>> all() {
        return KafkaFuture.allOf(futures.values().toArray(new KafkaFuture[0])).thenApply(
            nil -> {
                Map<String, ConsumerGroupDescription> descriptions = new HashMap<>(futures.size());
                futures.forEach((key, future) -> {
                    try {
                        descriptions.put(key.idValue, future.get());
                    } catch (InterruptedException | ExecutionException e) {
                        // This should be unreachable, since the KafkaFuture#allOf already ensured
                        // that all of the futures completed successfully.
                        throw new RuntimeException(e);
                    }
                });
                return descriptions;
            });
    }
```

skaundinya15 reviewed (approved):
Left a few very minor nit comments, other than that this looks great to me!

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/AdminApiDriver.java:
```diff
@@ -250,6 +252,13 @@ public void onFailure(
                 .collect(Collectors.toSet());
             retryLookup(keysToUnmap);
+
+        } else if (t instanceof NoBatchedFindCoordinatorsException) {
```
It feels a bit weird to handle a special case like this one in the driver. It is probably OK for the time being but we don't want to add more custom cases like this in the driver, I think. I wonder if we could delegate the decision to the handler. We could add an `handleUnsupportedVersionException` method to the `AdminApiHandler` for this purpose to delegate the decision. That method could basically return the keys to unmap and the keys to complete with the exception.

An alternative would be to rely on the the `handleUnsupportedVersionException` method in `Call`. The driver could also implement it and still delegate the decision to the handler. The advantage of using this method is that downgrade would not be counted as failures and thus does not count for the retries.

Have you considered something like this?

mimaison replied:
`AdminApiDriver` was still evolving rapidly while I was implementing this KIP so I went for the straighforward approach.

But I agree, it would be best to avoid this type of logic here. The goal would be to find a mechanism that works for all clients. @tombentley suggested an alternative option in https://github.com/apache/kafka/pull/10743#discussion_r649872433. 

I've not had the time to look into better alternatives yet.

dajac replied:
Should we file a Jira to not forget about improving this?

mimaison replied:
I opened https://issues.apache.org/jira/browse/KAFKA-13013

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/CoordinatorStrategy.java:
```diff
@@ -17,84 +17,158 @@
+
+        for (CoordinatorKey key : unrepresentableKeys) {
+            failedKeys.put(key, new InvalidGroupIdException("The given group id '" +
+                            key.idValue + "' cannot be represented in a request."));
```
nit: Indentation seems off here.

dajac commented on clients/src/main/resources/common/message/FindCoordinatorRequest.json:
```diff
@@ -23,12 +23,16 @@
   // Version 3 is the first flexible version.
-  "validVersions": "0-3",
+  //
+  // Version 4 adds CoordinatorKeys
```
nit: Could we add the KIP number here? I would also explain a bit more the change. The comment you have in the response is much better for instance.

dajac commented on clients/src/main/resources/common/message/FindCoordinatorResponse.json:
```diff
@@ -22,20 +22,33 @@
   // Version 3 is the first flexible version.
-  "validVersions": "0-3",
+  //
+  // Version 4 adds support for batching via Coordinators
```
nit: Could we add the KIP number here?

dajac reviewed (commented):
Thanks for the update. I left a few more questions/comments.

mimaison committed:
Address feedback

mimaison committed:
Mention KIP in json schema, fix indentation

dajac reviewed (approved):
LGTM. @mimaison Thanks for the PR! Could you file the follow-up Jiras that have been discussed in the comments? I have spotted two remaining ones: 1) one to address the error handling in the driver; 2) one to armonize the error handling in the handlers.

mimaison commented:
@dajac Thanks, I opened JIRAs for all follow up work items we identified.

mimaison committed:
Update tests to work with trunk

- Make KafkaAdminClientTest tests work with updated MockClient
- Update new ConsumerGroupServiceTest tests with new API

mimaison commented:
I've rebased on trunk to resolve conflicts and I've had to make small changes due to updates on trunk. All new changes are in https://github.com/apache/kafka/pull/10743/commits/ae09d742171c8a3a7d60587ce0502dce35945f73.

I'll let Jenkins run now and I'll merge later if there's no issues.
